### ISSUE TITLE
Finetuning Example

### ISSUE THREAD

**Issue by simonefrancia** [hanxiao/bert-as-service#76]
Hi, 
I am trying your code example (example5.py) to make a fine-tuning on my own mood dataset for a text classification task. I have already adapted the code and train starts correctly. 
After 5000 steps loss is descending but accuracy on validation is poor and constant at 42%. 

My doubt is that the fine-tuning is modifying weights of prediction task but not the input embeddings. When I used ELMo embedding for example, I set that embedding of input were trainable and classification task results are really better than this. 

Any suggestion about that? 

Thanks

**Comment by hanxiao** [comment]
my bad. I forgot to mask the sequence when doing reduce_mean, reduce_max and reduce_mean_max. this is fixed in #94 and now merged into master. please give it a try and may produce different result.

**Comment by simonefrancia** [comment]
I have tried new version but my training doesn't seem to be better. 
```
INFO:tensorflow:Calling model_fn.
INFO:tensorflow:Done calling model_fn.
INFO:tensorflow:Create CheckpointSaverHook.
INFO:tensorflow:Graph was finalized.
INFO:tensorflow:Running local_init_op.
INFO:tensorflow:Done running local_init_op.
INFO:tensorflow:Saving checkpoints for 0 into data/model/1206-081741/model.ckpt.
INFO:tensorflow:loss = 264.92365, step = 1
INFO:tensorflow:global_step/sec: 0.194579
INFO:tensorflow:loss = 142.13977, step = 101 (513.941 sec)
INFO:tensorflow:global_step/sec: 0.194874
INFO:tensorflow:loss = 152.33533, step = 201 (513.145 sec)
INFO:tensorflow:global_step/sec: 0.19485
INFO:tensorflow:loss = 159.28421, step = 301 (513.216 sec)
INFO:tensorflow:global_step/sec: 0.194599
INFO:tensorflow:loss = 152.00076, step = 401 (513.898 sec)
INFO:tensorflow:global_step/sec: 0.194962
INFO:tensorflow:loss = 144.81601, step = 501 (512.911 sec)
INFO:tensorflow:global_step/sec: 0.194442
INFO:tensorflow:loss = 134.33696, step = 601 (514.280 sec)
INFO:tensorflow:global_step/sec: 0.192623
INFO:tensorflow:loss = 144.52493, step = 701 (519.150 sec)
INFO:tensorflow:global_step/sec: 0.193265
INFO:tensorflow:loss = 144.21286, step = 801 (517.422 sec)
INFO:tensorflow:global_step/sec: 0.193592
INFO:tensorflow:loss = 159.57654, step = 901 (516.557 sec)
INFO:tensorflow:Saving checkpoints for 1000 into data/model/1206-081741/model.ckpt.
INFO:tensorflow:Calling model_fn.
INFO:tensorflow:Done calling model_fn.
INFO:tensorflow:Starting evaluation at 2018-12-06-09:43:44
INFO:tensorflow:Graph was finalized.
INFO:tensorflow:Restoring parameters from data/model/1206-081741/model.ckpt-1000
INFO:tensorflow:Running local_init_op.
INFO:tensorflow:Done running local_init_op.
INFO:tensorflow:Evaluation [10/100]
INFO:tensorflow:Evaluation [20/100]
INFO:tensorflow:Evaluation [30/100]
INFO:tensorflow:Evaluation [40/100]
INFO:tensorflow:Evaluation [50/100]
INFO:tensorflow:Evaluation [60/100]
INFO:tensorflow:Evaluation [70/100]
INFO:tensorflow:Evaluation [80/100]
INFO:tensorflow:Evaluation [90/100]
INFO:tensorflow:Evaluation [100/100]
INFO:tensorflow:Finished evaluation at 2018-12-06-09:53:55
INFO:tensorflow:Saving dict for global step 1000: accuracy = 0.433125, average_loss = 1.1821911, global_step = 1000, loss = 151.32047
INFO:tensorflow:Saving 'checkpoint_path' summary for global step 1000: data/model/1206-081741/model.ckpt-1000
INFO:tensorflow:global_step/sec: 0.0884598
INFO:tensorflow:loss = 143.34467, step = 1001 (1130.451 sec)
INFO:tensorflow:global_step/sec: 0.242282
INFO:tensorflow:loss = 152.80426, step = 1101 (412.750 sec)
INFO:tensorflow:global_step/sec: 0.170403
INFO:tensorflow:loss = 156.10487, step = 1201 (586.844 sec)
INFO:tensorflow:global_step/sec: 0.191289
INFO:tensorflow:loss = 180.17056, step = 1301 (522.766 sec)
INFO:tensorflow:global_step/sec: 0.189388
INFO:tensorflow:loss = 131.78491, step = 1401 (528.014 sec)
INFO:tensorflow:global_step/sec: 0.191926
INFO:tensorflow:loss = 159.46521, step = 1501 (521.031 sec)
INFO:tensorflow:global_step/sec: 0.189784
INFO:tensorflow:loss = 149.55505, step = 1601 (526.921 sec)
INFO:tensorflow:global_step/sec: 0.190483
INFO:tensorflow:loss = 171.7102, step = 1701 (524.980 sec)
INFO:tensorflow:global_step/sec: 0.190815
INFO:tensorflow:loss = 144.19461, step = 1801 (524.061 sec)
INFO:tensorflow:global_step/sec: 0.191387
INFO:tensorflow:loss = 155.31046, step = 1901 (522.512 sec)
INFO:tensorflow:Saving checkpoints for 2000 into data/model/1206-081741/model.ckpt.
INFO:tensorflow:Calling model_fn.
INFO:tensorflow:Done calling model_fn.
INFO:tensorflow:Starting evaluation at 2018-12-06-11:20:31
INFO:tensorflow:Graph was finalized.
INFO:tensorflow:Restoring parameters from data/model/1206-081741/model.ckpt-2000
INFO:tensorflow:Running local_init_op.
INFO:tensorflow:Done running local_init_op.
INFO:tensorflow:Evaluation [10/100]
INFO:tensorflow:Evaluation [20/100]
INFO:tensorflow:Evaluation [30/100]
INFO:tensorflow:Evaluation [40/100]
INFO:tensorflow:Evaluation [50/100]
INFO:tensorflow:Evaluation [60/100]
INFO:tensorflow:Evaluation [70/100]
INFO:tensorflow:Evaluation [80/100]
INFO:tensorflow:Evaluation [90/100]
INFO:tensorflow:Evaluation [100/100]
INFO:tensorflow:Finished evaluation at 2018-12-06-11:31:09
INFO:tensorflow:Saving dict for global step 2000: accuracy = 0.43398437, average_loss = 1.1828988, global_step = 2000, loss = 151.41104
INFO:tensorflow:Saving 'checkpoint_path' summary for global step 2000: data/model/1206-081741/model.ckpt-2000
INFO:tensorflow:global_step/sec: 0.0858941
INFO:tensorflow:loss = 160.1148, step = 2001 (1164.219 sec)
INFO:tensorflow:global_step/sec: 0.235182
INFO:tensorflow:loss = 158.74728, step = 2101 (425.207 sec)
```

If it can help,  I provide my code [finetuning_bert.zip] that is substantially the same to yours. 

[finetuning_bert.zip](https://github.com/hanxiao/bert-as-service/files/2652946/finetuning_bert.zip)

Thanks

**Comment by hanxiao** [comment]
fyi, serving a fine-tuned BERT model for feature extraction is fixed in #147 and available since 1.6.1. Please do 
```bash
pip install -U bert-serving-server bert-serving-client
```

More on this topics: https://github.com/hanxiao/bert-as-service#serving-a-fine-tuned-bert-model

**Comment by basque21** [comment]
Same problem here, validation accuracy not improving. Any help? and, @simonefrancia could you give me some tutorial or code where it is trained with Elmo embeddings? Are your results much more better? 
thanks!

**Comment by simonefrancia** [comment]
Hi @basque21,
I followed this guide and it's a good start for approaching ELMo:
```https://towardsdatascience.com/how-to-use-elmo-word-vectors-for-spam-classification-1891c0da8f1d```
